pkg/reconciler/ingress/resources/virtual_service.go

```diff
@@ -216,10 +216,16 @@ func makePortSelector(ios intstr.IntOrString) v1alpha3.PortSelector {
 	}
 }
 
-func makeVirtualServiceRoute(hosts sets.String, http *v1alpha1.HTTPIngressPath, gateways sets.String) *v1alpha3.HTTPRoute {
+func makeVirtualServiceRoute(hosts sets.String, http *v1alpha1.HTTPIngressPath, gateways map[v1alpha1.IngressVisibility]sets.String, visibility v1alpha1.IngressVisibility) *v1alpha3.HTTPRoute {
 	matches := []v1alpha3.HTTPMatchRequest{}
+	clusterDomainName := network.GetClusterDomainName()
 	for _, host := range hosts.List() {
-		matches = append(matches, makeMatch(host, http.Path, gateways))
+		g := gateways[visibility]
+		if strings.HasSuffix(host, clusterDomainName) {
+			// For local hostname, always use private gateway
```

Comment:
The private gateway may not be already existed. It is not a strict requirement, and users may not need it if they don't need to have cluster-local Routes.
So I think the logic here should be if there is cluster-local gateway, we bind local hostnames to cluster-local gateway. Otherwise, we fall back to public Gateway. WDYT?

Reply:
sure, we can check if the cluster-local gateway exists. If not, we would revert to the previous behavior.